File: src/gantry/runtime/pricing.py
"""Model pricing → USD, for the per-run cost ledger and budget brake.

Deliberately a small, stable table plus a conservative fallback: the exact
figure matters less than a deterministic, monotonic estimate the $ brake can
trust across resumes. Cached-input tokens are priced separately (cheap) so the
brake doesn't trip early once prompt caching engages. Prices are USD per 1M
tokens; override the table for your provider as needed.
"""

from __future__ import annotations

from dataclasses import dataclass

from gantry.runtime.llm import LLMUsage

_M = 1_000_000.0


@dataclass(frozen=True)
class ModelPrice:
    input_per_mtok: float
    output_per_mtok: float
    #: Cached-input read rate (typically ~0.1x input). Kept separate so the brake
    #: doesn't over-count re-read prefix tokens.
    cache_read_per_mtok: float


#: USD per 1M tokens. Public list prices (rounded); matched case-insensitively by
#: exact key first, then by the bare model slug. Extend for your providers.
_PRICING: dict[str, ModelPrice] = {
    "anthropic/claude-opus-4-8": ModelPrice(5.0, 25.0, 0.5),
    "anthropic/claude-sonnet-5": ModelPrice(3.0, 15.0, 0.3),
    "anthropic/claude-haiku-4-5": ModelPrice(1.0, 5.0, 0.1),
    "openrouter/qwen/qwen3-coder": ModelPrice(1.0, 3.0, 0.1),
    "openrouter/deepseek/deepseek-chat": ModelPrice(0.5, 1.5, 0.05),
    "openrouter/deepseek/deepseek-r1": ModelPrice(1.0, 3.0, 0.1),
}
#: Used for any model not in the table — conservative (errs toward tripping the
#: brake a little early rather than overrunning the budget).
_FALLBACK = ModelPrice(2.0, 6.0, 0.2)


def price_for(model: str) -> ModelPrice:
    m = model.lower()
    if m in _PRICING:
        return _PRICING[m]
    for key, price in _PRICING.items():
        if key in m or m.endswith(key.rsplit("/", 1)[-1]):
            return price
    return _FALLBACK


def cost_usd(usage: LLMUsage, model: str) -> float:
    """USD cost of one call's token usage. Cache-read tokens are billed at the
    cheap cached rate; the rest of the prompt at the input rate."""
    price = price_for(model)
    cached = max(0, usage.cache_read_tokens)
    fresh_input = max(0, usage.prompt_tokens - cached)
    return (
        fresh_input * price.input_per_mtok
        + cached * price.cache_read_per_mtok
        + usage.completion_tokens * price.output_per_mtok
    ) / _M

File: web/src/lib/pricing.ts
// Model pricing mirror of the backend `runtime/pricing.py`, so the dashboard can
// show a live dollar figure for the savings prompt-caching buys. Prices are USD
// per 1M tokens; kept deliberately small and matched the same way the backend
// does (exact key, then bare-slug contains) so the two never disagree on a cost.

import type { Task } from '../api/types'

export interface ModelPrice {
  input: number
  output: number
  cacheRead: number
}

const M = 1_000_000

const PRICING: Record<string, ModelPrice> = {
  'anthropic/claude-opus-4-8': { input: 5.0, output: 25.0, cacheRead: 0.5 },
  'anthropic/claude-sonnet-5': { input: 3.0, output: 15.0, cacheRead: 0.3 },
  'anthropic/claude-haiku-4-5': { input: 1.0, output: 5.0, cacheRead: 0.1 },
  'openrouter/qwen/qwen3-coder': { input: 1.0, output: 3.0, cacheRead: 0.1 },
  'openrouter/deepseek/deepseek-chat': { input: 0.5, output: 1.5, cacheRead: 0.05 },
  'openrouter/deepseek/deepseek-r1': { input: 1.0, output: 3.0, cacheRead: 0.1 },
}

const FALLBACK: ModelPrice = { input: 2.0, output: 6.0, cacheRead: 0.2 }

/** USD-per-1M price for a model slug, matched exactly then by bare slug — the
 * same order as the backend, so a cost shown here matches the ledger. */
export function priceFor(model: string): ModelPrice {
  const m = model.toLowerCase()
  if (m in PRICING) return PRICING[m]
  for (const [key, price] of Object.entries(PRICING)) {
    const bare = key.split('/').pop() ?? key
    if (m.includes(key) || m.endsWith(bare)) return price
  }
  return FALLBACK
}

function num(v: unknown): number {
  return typeof v === 'number' && Number.isFinite(v) && v > 0 ? v : 0
}

/** Dollars saved on ONE task by serving its cache-read tokens at the cheap cached
 * rate instead of fresh input: cacheRead x (input - cacheRead) per token. */
export function taskCacheSavingsUsd(task: Pick<Task, 'payload' | 'result'>): number {
  const cacheRead = num(task.result?.cache_read_tokens)
  if (cacheRead === 0) return 0
  const model = typeof task.payload.model === 'string' ? task.payload.model : ''
  const price = priceFor(model)
  return (cacheRead * (price.input - price.cacheRead)) / M
}

/** Total dollars prompt-caching saved across a run's whole tree — the run-level
 * "cost saver" figure. A still-running task with no result yet contributes 0, so
 * this reads as savings-so-far while live. */
export function runCacheSavingsUsd(tasks: Array<Pick<Task, 'payload' | 'result'>>): number {
  return tasks.reduce((sum, t) => sum + taskCacheSavingsUsd(t), 0)
}
